rfcs/2020-04-20-Optimizer-minimize.md

```diff
@@ -0,0 +1,308 @@
+# Easily Customizable `Optimizer.minimize`
+
+
+| Status        | In Revision       |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [234](https://github.com/tensorflow/community/pull/234) |
+| **Author(s)** | [omalleyt12@](https://github.com/omalleyt12) |
+| **Sponsor**   | apassos@, fchollet@, karmel@                 |
+| **Updated**   | 2020-04-20                                           |
+
+## Objective
+
+Create an `Optimizer` API that gives `Optimizer` subclasses full control of gradient updates. The API should ensure `Optimizer`s can be accessed via a unified API, and will not leak abstractions. Training loops should not be required to know the internal details of how the `Optimizer` chooses to:
+
+* Scale losses and gradients
+
+* Aggregate gradients
+
+* Clip gradients
+
+* etc
+
+We also need to ensure we maintain endpoints with maximum flexibility for those users who do want control over these items.
+
+By creating this API, it will enable users to write training loops that are interoperable with a wide range of Optimizers.
+
+Specific use cases considered:
+
+* Gradient clipping
+
+* Mixed precision
+
+* `Horovod`
+
+## Background
+
+During backpropagation, there are 6 possible actions that can be taken when starting from a loss Tensor and ending with a Variable update:
+
+(1) Transform the loss
+
+(2) Calculate the gradients
+
+(3) Transform the unaggregated (per-device) gradients
+
+(4) Aggregate the gradients (across devices)
+
+(5) Transform the aggregated gradients
+
+(6) Apply a variable update based on the gradients
+
+We currently have three Optimizer endpoints that start at different points in this process:
+
+* `Optimizer.minimize` - handles 1-6
+
+* `Optimizer.apply_gradients(..., experimental_aggregate_gradients=True)` - handles 4-6
+
+* `Optimizer.apply_gradients(..., experimental_aggregate_gradients=False)` - handles 6
+
+However, there is no easy way for Optimizer subclasses to support custom logic in these steps. This proposal suggests a refactoring of the Optimizer class to achieve these goals.
+
+
+## Motivation
+
+This section discusses the experience of supporting mixed-precision and Horovod in Keras’s built-in training logic (hereafter called Model.fit).
+
+Keras now allows users to write custom training logic for their `Model`s via overriding `Model.train_step`: [code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L538). The default implementation of this method is 8 lines long, and fully supports all types of `Model`s, `loss`es, `metric`s, etc that Keras supports. It attempts to serve as a reference that users can copy / paste to start writing their own training logic.
```

Comment:
I think the two line references here should probably not reference master branch: they no longer point to the code you want to (as far as I can tell)